aiohttp/helpers.py

```diff
@@ -132,7 +136,14 @@ def iscoroutinefunction(func: Any) -> bool:
 json_re = re.compile(r"(?:application/|[\w.-]+/[\w.+-]+?\+)json$", re.IGNORECASE)
 
 
-class BasicAuth(namedtuple("BasicAuth", ["login", "password", "encoding"])):
+class Auth:
+    """Authorization handler base class"""
+
+    async def __call__(self, request) -> None:
```

Comment:
`__call__()` is for normal callables, it'd be weird for it to be async. There's an `__await__()` that exists for awaitables. Maybe don't go for magic methods? It looks like you're reinventing a protocol pattern.

Also, it feels like for the use-case you described using an async generator interface could be a better fit then an async function.

Reply:
Thanks for the feedback @webknjaz. 

It might be useful to collect these high-level design considerations in the corresponding issue #6915. I was using this PR largely just to give an example what an implementation *could* look like.